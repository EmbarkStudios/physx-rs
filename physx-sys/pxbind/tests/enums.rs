fn gen_enums(which: &str, to_emit: &'static [&str]) -> anyhow::Result<String> {
    let ast = pxbind::get_parsed_ast(format!("tests/data/enum/{which}"))?;

    let mut consumer = pxbind::consumer::AstConsumer::default();
    consumer.consume(&ast)?;

    let generated = {
        let record_filter = |_rb: &pxbind::consumer::RecBinding<'_>| false;
        let enum_filter = |eb: &pxbind::consumer::EnumBinding<'_>| {
            to_emit.is_empty() || to_emit.iter().any(|te| *te == eb.name)
        };

        let generator = pxbind::generator::Generator {
            record_filter: Box::new(record_filter),
            enum_filter: Box::new(enum_filter),
            ..Default::default()
        };

        let mut buffer = Vec::new();
        generator.generate_rust_enums(&consumer, &mut buffer, 0)?;
        String::from_utf8(buffer)?
    };

    Ok(generated)
}

/// Verifies we can generate appropriate Rust enums for vanilla, i32 default C++ enums
#[test]
fn i32() {
    insta::assert_snapshot!(gen_enums("i32.h", &[]).unwrap());
}

/// Verifies we can generate appropriate Rust enums for C++ enums that have been
/// old school style forced to be u32
#[test]
fn u32() {
    insta::assert_snapshot!(gen_enums("u32.h", &[]).unwrap());
}

/// Verifies that enums not scoped to a struct can be parsed
#[test]
fn simple() {
    insta::assert_snapshot!(gen_enums("simple.h", &[]).unwrap());
}

/// Verifies that enums and their `PxFlags<>` wrappers can be properly bound
#[test]
fn flags() {
    insta::assert_snapshot!(gen_enums(
        "flags.h",
        &["PxActorCacheFlag", "PxArticulationCacheFlag", "PxShapeFlag"]
    )
    .unwrap());
}
