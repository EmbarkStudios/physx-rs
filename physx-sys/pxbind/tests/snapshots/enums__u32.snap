---
source: physx-sys/pxbind/tests/enums.rs
expression: "gen_enums(\"u32.h\", &[]).unwrap()"
---
#[derive(Copy, Clone)]
#[repr(u32)]
pub enum PxThreadPriority {
    /// High priority
    eHIGH = 0,
    /// Above Normal priority
    eABOVE_NORMAL = 1,
    /// Normal/default priority
    eNORMAL = 2,
    /// Below Normal priority
    eBELOW_NORMAL = 3,
    /// Low priority.
    eLOW = 4,
    eFORCE_DWORD = 4294967295,
}

/// Flags which affect the behavior of PxShapes.
#[derive(Copy, Clone)]
#[repr(u32)]
pub enum PxShapeFlag {
    /// The shape will partake in collision in the physical simulation.
    ///
    /// It is illegal to raise the eSIMULATION_SHAPE and eTRIGGER_SHAPE flags.
    /// In the event that one of these flags is already raised the sdk will reject any
    /// attempt to raise the other.  To raise the eSIMULATION_SHAPE first ensure that
    /// eTRIGGER_SHAPE is already lowered.
    ///
    /// This flag has no effect if simulation is disabled for the corresponding actor (see [`PxActorFlag::eDISABLE_SIMULATION`]).
    eSIMULATION_SHAPE = 1,
    /// The shape will partake in scene queries (ray casts, overlap tests, sweeps, ...).
    eSCENE_QUERY_SHAPE = 2,
    /// The shape is a trigger which can send reports whenever other shapes enter/leave its volume.
    ///
    /// Triangle meshes and heightfields can not be triggers. Shape creation will fail in these cases.
    ///
    /// Shapes marked as triggers do not collide with other objects. If an object should act both
    /// as a trigger shape and a collision shape then create a rigid body with two shapes, one being a
    /// trigger shape and the other a collision shape. 	It is illegal to raise the eTRIGGER_SHAPE and
    /// eSIMULATION_SHAPE flags on a single PxShape instance.  In the event that one of these flags is already
    /// raised the sdk will reject any attempt to raise the other.  To raise the eTRIGGER_SHAPE flag first
    /// ensure that eSIMULATION_SHAPE flag is already lowered.
    ///
    /// Trigger shapes will no longer send notification events for interactions with other trigger shapes.
    ///
    /// Shapes marked as triggers are allowed to participate in scene queries, provided the eSCENE_QUERY_SHAPE flag is set.
    ///
    /// This flag has no effect if simulation is disabled for the corresponding actor (see [`PxActorFlag::eDISABLE_SIMULATION`]).
    eTRIGGER_SHAPE = 4,
    /// Enable debug renderer for this shape
    eVISUALIZATION = 8,
}

